### ISSUE TITLE
File-like objects are closed after uploading

### ISSUE THREAD

**Issue by JordonPhillips** [boto/s3transfer#80]
Source boto/boto3#929

This is because we wrap them with the `InterruptReader` which [closes](https://github.com/boto/s3transfer/blob/develop/s3transfer/upload.py#L100) them on exit.

**Comment by kyleknap** [comment]
Yeah looks like it will only happen for the non-multipart upload case because the provided file object is used. Will probably have to figure out a way to safeguard against accidentally closing the file.

**Comment by ramurti** [comment]
Any workaround to prevent the file being closed? I'm using this function with unnamed temporary files, so I can't simply reopen the file. I'd like to do something else with this file (for example, rotate it to local disk) in case upload_fileobj fails, but it closes the file.... (sigh)

**Comment by micahfay** [comment]
The workaround by leonsmith here worked for me- https://github.com/matthewwithanm/django-imagekit/issues/391

**Comment by ofassley** [comment]
Same issue here.  I'm also using Imagekit.

```
self = <ImageCacheFile: CACHE/avatars/a0cbd7767bf54e649f4952e8c3407418/9dd51a8dd80342ae7576d1c376b51e43.jpg>

    def _generate(self):
        # Generate the file
        content = generate(self.generator)

        actual_name = self.storage.save(self.name, content)

        # We're going to reuse the generated file, so we need to reset the pointer.
>       content.seek(0)
E       ValueError: I/O operation on closed file.

actual_name = 'CACHE/avatars/a0cbd7767bf54e649f4952e8c3407418/9dd51a8dd80342ae7576d1c376b51e43.jpg'
content    = <File: None>
self       = <ImageCacheFile: CACHE/avatars/a0cbd7767bf54e649f4952e8c3407418/9dd51a8dd80342ae7576d1c376b51e43.jpg>

../../../.virtualenvs/myapp/lib/python3.5/site-packages/imagekit/cachefiles/__init__.py:102: ValueError
```

**Comment by mgrdcm** [comment]
I can't tell - is this a bug?  I'll be happy to take a stab at a PR if it is, and if nobody else wants to handle it.  I'm also suffering from the `django-imagekit` issue.

**Comment by micahfay** [comment]
I've stopped using django-storages and moved to this for django-s3 stuff- lighter weight, and seems to be better supported :) https://github.com/etianen/django-s3-storage switching was easy even with existing data.

**Comment by joshma** [comment]
@kyleknap when you get a chance, could you chime in on whether calling `self._fileobj.close()` in `InterruptReader` is necessary? We're running into the same issue, and it's particularly difficult because we're accepting a few kinds of file-like objects so we can't just re-open it later.

**Comment by freemanlevinetang** [comment]
Ran into this issue recently while attempting to generate images, upload them to s3 for backup, and then attach them to an email.
Workaround was to pull the file back down from s3 after uploading, which seems pretty inefficient.

**Comment by leogzyl** [comment]
I'm using the following workaround:
```
from io import BufferedReader

class NonCloseableBufferedReader(BufferedReader):
    def close(self):
        pass

buffer = NonCloseableBufferedReader(myfile)
client.upload_fileobj(Fileobj=buffer, Bucket=bucket, Key=name)
buffer.detach()
assert not myfile.closed
```

**Comment by zachliu** [comment]
@mgrdcm I think it is a bug. The control should be relinquished to the end user instead of closing it prematurely.

**Comment by zachliu** [comment]
@leogzyl A near perfect workaround except `mypy` complains if I pass in `tempfile.TemporaryFile()` object:
```
error: Argument 1 to "NonCloseableBufferedReader" has incompatible type "IO[Any]"; expected "RawIOBase"
```
Any solutions other than `# type: ignore`?

**Comment by ncvc** [comment]
@zachliu I added an init method with a cast to work around the mypy issue:
```
class NonCloseableBufferedReader(BufferedReader):
    def __init__(self, raw: IO[bytes], *args: Any, **kwargs: Any):
        super().__init__(cast(RawIOBase, raw), *args, **kwargs)

    def close(self) -> None:
        self.flush()

NonCloseableBufferedReader(myfile)  # No mypy errors
```

**Comment by Jihadik** [comment]
3 years passed... Can it be either fixed or declared to be not a bug, so other libraries should build workarounds for it?

**Comment by jannotti** [comment]
I mean, it's clearly a bug, right?  You can't specify that a method maybe, sometimes, occasionally closes your files.

**Comment by arthurlogilab** [comment]
We're affected by this too. The NonCloseableBufferedReader workaround seems to do the trick, but hopefully we can remove it once this is fixed.

**Comment by rooom13** [comment]
The issue still appears on March 2020, @leogzyl 's workarounds solves the issue

**Comment by zivkovic** [comment]
There is many of us with this issue, something should be done with it.

**Comment by diego200052** [comment]
I got this same issue on April 2021 using boto3 and upload_fileobj() function.

**Comment by jvanasco** [comment]
This behavior is not documented.  This behavior is not expected.  The first version of boto did not implement this behavior.  I've been using Python for over 20 years and have never seen a library close a file in this context.  There is no reason why this function should close the file.  

The only thing more ridiculous than this bug existing, is that it has not been fixed in 4.5 years.  If keeping the file open causes an issue, then the library has serious underlying problems.

**Comment by iamahuman** [comment]
Suspected regression commit: boto/s3transfer@2f3d12cba458d7f90e583c754e317e479c49c8c9

Diff: https://github.com/boto/s3transfer/commit/2f3d12cba458d7f90e583c754e317e479c49c8c9#diff-fb5195c2216d4f5a452b31d518beae30b39667b0655d8a8fd1663a183575aa0aL167-R165

```diff
-        client.put_object(Bucket=bucket, Key=key, Body=body, **extra_args)
+        with osutil.open_file_chunk_reader(
+                fileobj, 0, size, progress_callbacks) as body:
+            client.put_object(Bucket=bucket, Key=key, Body=body, **extra_args)
```

**Comment by saranglakare** [comment]
5 years anniversary of this issue and it's still open! I just wasted a couple of hours figuring out my problem and it came to this bug! Why isn't anyone working on this?

**Comment by mgrdcm** [comment]
> 5 years anniversary of this issue and it's still open! I just wasted a couple of hours figuring out my problem and it came to this bug! Why isn't anyone working on this?

It's tradition now.

**Comment by yswtrue** [comment]
Same here

**Comment by antoniosarosi** [comment]
@saranglakare Almost 7 years and still counting :joy:, just run into this issue today.

**Comment by sangshuduo** [comment]
I encountered this bug on May 3, 2024.

**Comment by 1mamute** [comment]
Kinda disappointing this issue is almost 8 years old and not even a single mention in the documentation was made.

Here's another hacky workaround:
```python
def patched_close():
    print("good try boto3")

file = BytesIO()

# Save the original close function
original_close = file.close

# Patch the close function with ours
file.close = patched_close

client.upload_fileobj(Fileobj=file, Bucket=bucket, Key=name)

# Unpatch it
file.close = original_close
```

**Comment by aswanthabam** [comment]
I encountered this issue on **May 10, 2025**, and I’m surprised it’s still not resolved. I believe adding a parameter to control whether the file should be closed after the upload completes would help support more flexible workflows. Other libraries that rely on `s3transfer` could also make use of this flag as needed.

After reviewing the code, it looks like the change could be implemented by:

1. Adding a `close_file` parameter to `TransferManager.upload`.
2. Passing that through `CallArgs`, and into `TransferCoordinator`.
3. Using the flag in `InterruptReader.close()` to either close or flush the file.

Here’s a rough implementation outline:

### `manager.py`
```python
class TransferManager:
    ...
    def upload(self, fileobj, bucket, key, extra_args=None, subscribers=None, close_file=True):
        ...
        call_args = CallArgs(
            ...,
            close_file=close_file
        )
        return self._submit_transfer(
            call_args, UploadSubmissionTask, extra_main_kwargs
        )

    def _get_future_with_components(self, call_args):
        transfer_id = self._id_counter
        transfer_coordinator = TransferCoordinator(
            transfer_id=transfer_id,
            close_file=call_args.close_file
        )
````

### `futures.py`

```python
class TransferCoordinator:
    def __init__(self, transfer_id=None, close_file=True):
        self.transfer_id = transfer_id
        self.close_file = close_file
```

### `upload.py`

```python
class InterruptReader:
    ...
    def close(self):
        if self.transfer_coordinator.close_file:
            self._fileobj.close()
        else:
            self._fileobj.flush()
```

This can also be supported in **[boto3](https://github.com/boto/boto3)** by passing `close_file` through `upload_fileobj`.

If this approach sounds reasonable, I’d be happy to help with a PR. Let me know your thoughts! @JordonPhillips
